File: julia/perf.jl
# This file was formerly a part of Julia. License is MIT: https://julialang.org/license

using LinearAlgebra: tr
using Printf: @printf
using Random: seed!, rand, randn
using Statistics: std, mean
using Test: @test

const NITER = 5

seed!(1776)

macro timeit(ex, name)
    quote
        t = zeros(NITER)
        for i = 0:NITER
            e = 1000 * (@elapsed $(esc(ex)))
            if i > 0
                t[i] = e  # i == 0 is JIT warmup
            end
        end
        @printf "julia,%s,%f\n" $(esc(name)) minimum(t)
        GC.gc()
    end
end

## recursive fib ##

fib(n) = n < 2 ? n : fib(n - 1) + fib(n - 2)

@test fib(20) == 6765
@timeit fib(20) "recursion_fibonacci"

## parse integer ##

function parseintperf(t)
    local n
    for i = 1:t
        n = rand(UInt32)
        s = string(n, base=16)
        m = parse(UInt32, s, base=16)
        @assert m == n
    end
    return n
end

@timeit parseintperf(1000) "parse_integers"

## mandelbrot set: complex arithmetic and comprehensions ##

function mandel(z)
    c = z
    maxiter = 80
    for n = 1:maxiter
        if real(z)*real(z) + imag(z)*imag(z) > 4
            return n - 1
        end
        z = z^2 + c
    end
    return maxiter
end

mandelperf() = [mandel(complex(r, i)) for i = -1.0:0.1:1.0, r = -2.0:0.1:0.5]
@test sum(mandelperf()) == 14791
@timeit mandelperf() "userfunc_mandelbrot"

## numeric vector sort ##

function qsort!(a, lo, hi)
    i, j = lo, hi
    while i < hi
        pivot = a[(lo+hi)>>>1]
        while i <= j
            while a[i] < pivot; i += 1; end
            while a[j] > pivot; j -= 1; end
            if i <= j
                a[i], a[j] = a[j], a[i]
                i, j = i + 1, j - 1
            end
        end
        if lo < j; qsort!(a, lo, j); end
        lo, j = i, hi
    end
    return a
end

sortperf(n) = qsort!(rand(n), 1, n)
@test issorted(sortperf(5000))
@timeit sortperf(5000) "recursion_quicksort"

## slow pi series ##

const _pisum_vol = Ref(0.0)

function pisum()
    for j = 1:500
        s = 0.0
        for k = 1:10000
            s += 1.0 / (k * k)
        end
        _pisum_vol[] = s
    end
    return _pisum_vol[]
end

@test abs(pisum() - 1.644834071848065) < 1e-12
@timeit pisum() "iteration_pi_sum"

## random matrix statistics ##

function randmatstat(t)
    n = 5
    v = zeros(t)
    w = zeros(t)
    for i = 1:t
        a = randn(n, n)
        b = randn(n, n)
        c = randn(n, n)
        d = randn(n, n)
        P = [a b c d]
        Q = [a b; c d]
        v[i] = tr((P'P)^4)
        w[i] = tr((Q'Q)^4)
    end
    return (std(v) / mean(v), std(w) / mean(w))
end

(s1, s2) = randmatstat(1000)
@test 0.5 < s1 < 1.0 && 0.5 < s2 < 1.0
@timeit randmatstat(1000) "matrix_statistics"

## largish random number gen & matmul ##

@timeit rand(1000, 1000) * rand(1000, 1000) "matrix_multiply"

## printfd ##

if Sys.isunix()
    function printfd(n)
        open("/dev/null", "w") do io
            for i = 1:n
                @printf(io, "%d %d\n", i, i + 1)
            end
        end
    end

    printfd(1)
    @timeit printfd(100000) "print_to_file"
end

File: scala/src/main/scala/perf.scala
// This file was formerly a part of Julia. License is MIT: https://julialang.org/license

import java.io.{BufferedOutputStream, FileOutputStream, PrintStream}
import java.util.Random

import breeze.linalg._
import breeze.stats.distributions.Gaussian

object Perf {
  val NITER = 5

  def printPerf(name: String, t: Long): Unit = {
    printf("scala,%s,%.6f\n", name, t / 1e6)
  }

  // ---------------------------------------------------------------------------
  // Fibonacci
  // ---------------------------------------------------------------------------

  def fib(n: Int): Int = if (n < 2) n else fib(n - 1) + fib(n - 2)

  // ---------------------------------------------------------------------------
  // Mandelbrot
  // ---------------------------------------------------------------------------

  def mandel(zr0: Double, zi0: Double): Int = {
    val maxiter = 80
    var zr = zr0
    var zi = zi0
    var n = 0
    while (n < maxiter) {
      if (zr * zr + zi * zi > 4.0) return n
      val zrNew = zr * zr - zi * zi + zr0
      val ziNew = 2.0 * zr * zi + zi0
      zr = zrNew
      zi = ziNew
      n += 1
    }
    maxiter
  }

  def mandelperf(): Array[Int] = {
    val M = new Array[Int](21 * 26)
    var i = 0
    while (i < 21) {
      var j = 0
      while (j < 26) {
        M(26 * i + j) = mandel((j - 20) / 10.0, (i - 10) / 10.0)
        j += 1
      }
      i += 1
    }
    M
  }

  // ---------------------------------------------------------------------------
  // Quicksort
  // ---------------------------------------------------------------------------

  def quicksort(a: Array[Double], lo0: Int, hi: Int): Unit = {
    var i = lo0
    var j = hi
    var lo = lo0
    while (i < hi) {
      val pivot = a((lo + hi) / 2)
      while (i <= j) {
        while (a(i) < pivot) i += 1
        while (a(j) > pivot) j -= 1
        if (i <= j) {
          val tmp = a(i)
          a(i) = a(j)
          a(j) = tmp
          i += 1
          j -= 1
        }
      }
      if (lo < j) quicksort(a, lo, j)
      lo = i
      j = hi
    }
  }

  // ---------------------------------------------------------------------------
  // Pi sum
  // ---------------------------------------------------------------------------

  def pisum(): Double = {
    var sum = 0.0
    var j = 0
    while (j < 500) {
      sum = 0.0
      var k = 1
      while (k <= 10000) {
        sum += 1.0 / (k * k)
        k += 1
      }
      j += 1
    }
    sum
  }

  // ---------------------------------------------------------------------------
  // Print to file
  // ---------------------------------------------------------------------------

  def printfd(n: Int): Unit = {
    val ps = new PrintStream(new BufferedOutputStream(new FileOutputStream("/dev/null")))
    try {
      var i = 0
      while (i < n) {
        ps.println(s"$i ${i + 1}")
        i += 1
      }
    } finally {
      ps.close()
    }
  }

  // ---------------------------------------------------------------------------
  // Random matrix statistics (Breeze / BLAS)
  // ---------------------------------------------------------------------------

  def randmatstat(t: Int): (Double, Double) = {
    val n = 5
    val v = new Array[Double](t)
    val w = new Array[Double](t)
    val g = Gaussian(0, 1)

    for (i <- 0 until t) {
      val a = DenseMatrix.rand(n, n, g)
      val b = DenseMatrix.rand(n, n, g)
      val c = DenseMatrix.rand(n, n, g)
      val d = DenseMatrix.rand(n, n, g)

      val P = DenseMatrix.horzcat(a, b, c, d)
      val Q = DenseMatrix.vertcat(DenseMatrix.horzcat(a, b), DenseMatrix.horzcat(c, d))

      val V = P.t * P
      val W = Q.t * Q

      v(i) = trace(V * V * V * V)
      w(i) = trace(W * W * W * W)
    }

    def stdev(arr: Array[Double]): Double = {
      val mean = arr.sum / arr.length
      val variance = arr.map(x => (x - mean) * (x - mean)).sum / (arr.length - 1)
      math.sqrt(variance)
    }

    def mean(arr: Array[Double]): Double = arr.sum / arr.length

    (stdev(v) / mean(v), stdev(w) / mean(w))
  }

  // ---------------------------------------------------------------------------
  // Random matrix multiply (Breeze / BLAS)
  // ---------------------------------------------------------------------------

  def randmatmul(n: Int): DenseMatrix[Double] = {
    val a = DenseMatrix.rand[Double](n, n)
    val b = DenseMatrix.rand[Double](n, n)
    a * b
  }

  // ---------------------------------------------------------------------------
  // Main
  // ---------------------------------------------------------------------------

  def main(args: Array[String]): Unit = {
    val rand = new Random(0)
    var tmin = 0L
    var t = 0L

    // recursion_fibonacci
    assert(fib(20) == 6765)
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      fib(20)
      tmin = math.min(tmin, System.nanoTime() - t)
    }
    printPerf("recursion_fibonacci", tmin)

    // parse_integers
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      for (_ <- 0 until 1000) {
        val n = rand.nextInt(Integer.MAX_VALUE)
        val s = Integer.toHexString(n)
        val m = Integer.parseInt(s, 16)
        assert(m == n)
      }
      tmin = math.min(tmin, System.nanoTime() - t)
    }
    printPerf("parse_integers", tmin)

    // userfunc_mandelbrot
    {
      val M = mandelperf()
      val s = M.sum
      assert(s == 14791, s"mandel sum was $s, expected 14791")
    }
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      mandelperf()
      tmin = math.min(tmin, System.nanoTime() - t)
    }
    printPerf("userfunc_mandelbrot", tmin)

    // recursion_quicksort
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      val d = new Array[Double](5000)
      var j = 5000
      while (j > 0) {
        j -= 1
        d(j) = rand.nextDouble()
      }
      quicksort(d, 0, 4999)
      tmin = math.min(tmin, System.nanoTime() - t)
    }
    printPerf("recursion_quicksort", tmin)

    // iteration_pi_sum
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      val pi = pisum()
      tmin = math.min(tmin, System.nanoTime() - t)
      assert(math.abs(pi - 1.644834071848065) < 1e-12)
    }
    printPerf("iteration_pi_sum", tmin)

    // matrix_statistics
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      randmatstat(1000)
      tmin = math.min(tmin, System.nanoTime() - t)
    }
    printPerf("matrix_statistics", tmin)

    // matrix_multiply
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      val m = randmatmul(1000)
      assert(0 <= m(0, 0))
      tmin = math.min(tmin, System.nanoTime() - t)
    }
    printPerf("matrix_multiply", tmin)

    // print_to_file
    tmin = Long.MaxValue
    for (_ <- 0 until NITER) {
      t = System.nanoTime()
      printfd(100000)
      tmin = math.min(tmin, System.nanoTime() - t)
    }
    printPerf("print_to_file", tmin)
  }
}
